sacred/exp_runner.py

```diff
@@ -0,0 +1,335 @@
+
+            # ex.observers.clear()
+            # ex.observers = [o for o in ex.observers if not isinstance(o, MongoObserver)]
+            assert isinstance(ex.observers[0], MongoObserver)
+            ex.observers[0] = observer
+            # Done:
+        except BaseException as e:
+            observer.requeue(run)
+            raise e
+        else:
+            result = ex.run_command(run['command'], run['config'])
+            logger.info('job finished: %s' % str(run['_id']))
+
+
+@contextmanager
+def working_dir(dir):
+    cwd = os.getcwd()
+    os.chdir(dir)
+    yield
+    os.chdir(cwd)
+
+
+def exec_import(cmd_module, cmd):
+    worker_fn = None
+    exec('import %s; worker_fn = %s' % (cmd_module, cmd))
+    return worker_fn
+
+
+def as_mongo_str(s):
+    if s.startswith('mongodb://'):
+        return s
+    else:
+        return 'mongodb://%s' % s
+
+
+def main_worker_helper(options, args):
+    N = int(options.max_jobs)
+    if options.last_job_timeout is not None:
+        last_job_timeout = time.time() + float(options.last_job_timeout)
+    else:
+        last_job_timeout = None
+
+    def sighandler_shutdown(signum, frame):
+        logger.info('Caught signal %i, shutting down.' % signum)
+        raise Shutdown(signum)
+
+    def sighandler_wait_quit(signum, frame):
+        logger.info('Caught signal %i, shutting down.' % signum)
+        raise WaitQuit(signum)
+
+    signal.signal(signal.SIGINT, sighandler_shutdown)
+    signal.signal(signal.SIGHUP, sighandler_shutdown)
+    signal.signal(signal.SIGTERM, sighandler_shutdown)
+    signal.signal(signal.SIGUSR1, sighandler_wait_quit)
+
+    if N > 1:
+        proc = None
+        cons_errs = 0
+        if last_job_timeout and time.time() > last_job_timeout:
+            logger.info("Exiting due to last_job_timeout")
+            return
+
+        while N and cons_errs < int(options.max_consecutive_failures):
+            try:
+                # recursive Popen, dropping N from the argv
+                # By using another process to run this job
+                # we protect ourselves from memory leaks, bad cleanup
+                # and other annoying details.
+                # The tradeoff is that a large dataset must be reloaded once for
+                # each subprocess.
+                sub_argv = [sys.argv[0],
+                            '--poll-interval=%s' % options.poll_interval,
+                            '--max-jobs=1',
+                            '--mongo=%s' % options.mongo,
+                            '--db_name=%s' % options.db_name,
+                            '--reserve-timeout=%s' % options.reserve_timeout]
+                if options.workdir is not None:
+                    sub_argv.append('--workdir=%s' % options.workdir)
+                if options.ssh is not None:
+                    sub_argv.append('--ssh=%s' % options.ssh)
+                if options.exp_key is not None:
+                    sub_argv.append('--exp-key=%s' % options.exp_key)
```

Comment:
Why pass through arguments individually? Couldn't you just search and replace the `max_jobs` one?